apps/noodle/librapi2/branches/0.9.3-3guada1/tools/man/pstatus.1
.\" $Id: pstatus.1 2348 2006-04-05 14:58:44Z twogood $
.TH "PSTATUS" "1" "November 2002" "The SynCE project" "http://synce.sourceforge.net/"
.SH NAME
pstatus \- show device status

.SH SYNOPSIS
\fBpstatus\fR [-d \fILEVEL\fR] [-p \fIDEVPATH\fR] [-h]\fR

.SH "DESCRIPTION"

.PP
\fBpstatus\fR displays information about a device connected through SynCE.

.SH "OPTIONS"
.TP
-d \fILEVEL\fR
Set debug log level:
.IP
0 - No logging (default)
.IP
1 - Errors only
.IP
2 - Errors and warnings
.IP
3 - Everything

.TP
-p \fIDEVPATH\fR
Use the device at the given path, instead of the default.

.TP
-h
Display help message.




.SH "AUTHOR"
.PP
David Eriksson <twogood@users.sourceforge.net>.
.SH "SEE ALSO"
synce(1) prun(1)


Changelog
BETA3 (n/a)
===========

PERSONALIZACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Cerrada lista prácticamente definitiva de aplicaciones
- Incluidas nuevas aplicaciones:
  - lsr
  - mozilla-acroread
  - dasher
  - gFTP (BTS 1087)
  - xchat (BTS 1086)
  - bluefish (BTS 1085)
  - d4x (BTS 1084)
  - soundtracker (BTS 1063)
  - gnome-games
  - firestarter
  - streamtuner
  - aMSN (v. 0.96)
  - gdebi
  - audacity (BTS 1059)
- Incluidas aplicaciones de consola (del spec): mc, ncftp, minicom, vim, ssh,
wget, rsync, dillo, sqlite, nmap, gsm-utils, brltty, joe
- Mejoras en el sistema de suplementos: comprobación de si el suplemento se
instala correctamente, comprobación de dependencias,...
- Adaptación de "contenido para Windows" de la v3 a la v4: pendiente de
terminar de depurar (BTS 1044)
- Solucionado problema de acceso a unidades SAMBA de la red: no se refrescaban
correctamente las unidades de red a través de nautilus
- Incluidos ficheros de ejemplo, adaptando los de la v3
- Pruebas de hibernación y gestión de energía en equipos portátiles correctas
- Añadido asistente para instalación de codecs de audio/vídeo
- Homogeneizadas entradas del menú: descripción (nombre aplicación)
- Implementado e incluido asistente para conexión a Internet
- Incluido "watermain"
- Traducido "gxine" (al final, descartada de la distribución base)
- Incluida imagen de fondo de Grub
- Personalizada página de prueba de impresión
- Configuración adecuada de "rhythmbox": inclusión de algunas radios
nacionales (tal y como se indica en el "spec")
- Configuración adecuada de "liferea": inclusión de algunos periódicos
nacionales (tal y como se indica en el "spec")
- Las aplicaciones en "Herramientas de Sistema" pasan a "Sistema ->
Administración"
- Configuración adecuada de "amule": navegador y reproductor de vídeo por
defecto (BTS 1017)


INSTALADOR LIVE (LIVE-INSTALLER)
--------------------------------

- Añadido visualizador de particiones de disco en selección de puntos de
montaje
- Mejorado (más usable y sencillo) el resumen final del instalador (BTS 955)
- Corregido problema de instalación de paquetes de soporte de inglés al final
de la instalación (ya no los instala)
- Corregido error de instalación cuando existen múltiples discos duros (BTS
965 y 1126)
- Mejorado sistema de redimensionado: permitía disminuir el tamaño por debajo
del mínimo necesario para la instalación
- Mejorado opción de utilizar espacio libre: permitía usar espacio sin
particionar demasiado pequeño
- Añadido soporte para instalación del perfil de accesibilidad con el que se
arranca en live al usuario que se crea en el sistema instalado
- Mejoradas descripciones de opciones de particionado
- Corregidos errores de unidades montadas tras usar "DiskPreview"
- Establecidos valores mínimos de instalación: 3 GB (/), 500 MB (/home), 64 MB
(swap)
- Análisis de soporte de accesibilidad de "gparted": habría que hacer un "fork"
- Personalizado botón de instalar incluyendo nuevo logo

INSTALADOR NO LIVE (DI-INSTALLER)
---------------------------------

- Corregidos algunos problemas leves de traducción

ACCESIBILIDAD
-------------

- Análisis y repaso de perfiles de accesibilidad
- Validación de funcionamiento de los perfiles de accesibilidad

NOODLE
------

- Implementado e incluido actor de Hermes para gestión de PDAs
- Finalizada integración en Guadalinex: paquetes guadalinex-noodle
- Análisis de la tecnología "opensync" para sincronización a través de
"syncml" (eg: symbian); probadas diferentes versiones (0.18 (dapper), 0.19,
020svn): descartado por considerarse poco estable aún.
- Corregidos algunos errores.

SISTEMA DE GENERACIÓN (GENERATION SYSTEM)
-----------------------------------------

- Actualizados paquetes relevantes de "-updates", "-security" y "-backports"
- Mejorada la estructura de metapaquetes de Guadalinex para hacer más flexible
la desinstalación de paquetes sin afectar a las configuraciones del resto

E-ADMIN
-------

- Incluido sistema de instalación de certificado raíz de la FNMT:
  - Cuando se detecta un certificado de usuario y se va a instalar en Firefox,
  se pide permiso al usuario para la instalación del certificado raíz,
  pidiendo que acepte la licencia.
  - Hecho lo anterior, se instala el certificado raíz en el sistema
- Modificado sistema de creación de usuarios para añadir a los nuevos usuarios
al grupo "scard"
- Añadidas librerías "jss" para acceso a diferentes servicios Java de la JA
- Análisis de firma en Evolution con DNIe: descartado (no es posible firmar
con el DNIe)


BETA2 (12/01/07)
================

PERSONALICACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Incluido "ndiswrapper" para soporte de algunas tarjetas wifi
- Añadidos iconos en el escritorio: "carpeta personal" y "equipo".
- Repaso e inclusión en Guadalinex de todos los paquetes interesantes que
incluye Ubuntu Edgy y que no estaban en Guadalinex: evolution-exchange,...
- Añadidos repositorios oficiales de Guadalinex v4
- Implementado interfaz de usuario para suplementos de manera que se pueda
utilizar sin hermes
- Compartición del directorio "Compartido" para permitir a los usuarios
compartir ficheros a través de SAMBA
- Regenerados "grub" y "grub-installer" para configurar adecuadamente el
nombre de la distribución ("Ubuntu" -> "Guadalinex")
- Comprobación de detección de entradas de GRUB para otros sistemas
operativos tras instalación
- Implementación de nueva funcionalidad para gnome-system-tools: soporte de
perfiles de accesibilidad para los usuarios que se crean
- Establecido número de escritorios virtuales a 4
- Añadido "alacarte" para edición de menús
- Añadido "camorama" para uso de webcams
- Añadido "lemurae" para acceso al diccionario de la RAE
- Traducido mensaje inicial al abrir una consola
- Personalizado el icono de salida de sesión
- Deshabilitado servidor SSH por defecto

apps/noodle/librapi2/trunk/tools/pstatus.c
/* $Id: pstatus.c 2348 2006-04-05 14:58:44Z twogood $ */
#include <rapi.h>
#include <synce_log.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

char* devpath = NULL;

static void show_usage(const char* name)
{
  fprintf(stderr,
      "Syntax:\n"
      "\n"
      "\t%s [-d LEVEL] [-p DEVPATH] [-h]\n"
      "\n"
      "\t-d LEVEL  Set debug log level\n"
      "\t              0 - No logging (default)\n"
      "\t              1 - Errors only\n"
      "\t              2 - Errors and warnings\n"
      "\t              3 - Everything\n"
      "\t-h        Show this help message\n"
      "\t-p DEVPATH   Device path\n",    
      name);
}

static bool handle_parameters(int argc, char** argv)
{
  int c;
  int log_level = SYNCE_LOG_LEVEL_LOWEST;

  while ((c = getopt(argc, argv, "d:hp:")) != -1)
  {
    switch (c)
    {
      case 'd':
        log_level = atoi(optarg);
        break;

      case 'p':
        devpath = optarg;
        break;

      case 'h':
      default:
        show_usage(argv[0]);
        return false;
    }
  }

  synce_log_set_level(log_level);

  return true;
}

static const char* version_string(CEOSVERSIONINFO* version)
{
  const char* result = "Unknown";

  if (version->dwMajorVersion == 4)
  {
    if (version->dwMinorVersion == 20 && version->dwBuildNumber == 1081)
      result = "Ozone: Pocket PC 2003 (?)";
    else if (version->dwMinorVersion == 21 && version->dwBuildNumber == 1088)
      result = "Microsoft Windows Mobile 2003 Pocket PC Phone Edition";
  } 
  else if (version->dwMajorVersion == 3 &&
      version->dwMinorVersion == 0)
  {
    switch (version->dwBuildNumber)
    {
      case 9348:  result = "Rapier: Pocket PC"; break;
      case 11171: result = "Merlin: Pocket PC 2002"; break;

      /* 
       * From:     Jonathan McDowell
       * To:       SynCE-Devel
       * Subject:  Re: [Synce-devel] Smartphone & installing CABs.
       * Date: 	   Mon, 26 May 2003 19:12:10 +0100  (20:12 CEST)
       */
      case 12255: result = "Stinger: Smart Phone 2002"; break;

      /* My Qtek 7070 */
      case 13121: result = "Stinger: Smart Phone 2002"; break;
    }
  }
  else if (version->dwMajorVersion == 2 &&
      version->dwMinorVersion == 1)
  {
    result = 
      "Gryphon: Windows CE for P/PC V1 (Palm-size PC)"
      " / "
      "Apollo: Windows CE for A/PC V1 (Auto PC)";
  }

  return result;
}

#define PROCESSOR_ARCHITECTURE_COUNT 8

static const char* architecture[] = {
  "Intel",
  "MIPS",
  "Alpha",
  "PPC",
  "SHX",
  "ARM",
  "IA64",
  "ALPHA64"
};

static const char* processor(int n)
{
  const char* result;

  switch (n)
  {
    case PROCESSOR_STRONGARM:    result = "StrongARM";  break;
    case PROCESSOR_MIPS_R4000:   result = "MIPS R4000"; break;
    case PROCESSOR_HITACHI_SH3:  result = "SH3";        break;

    default: result = "Unknown"; break;
  }

  return result;
}

#if 0
static void print_flag(unsigned flags, unsigned flag, const char*name, bool* first)
{
  if (flags & flag)
  {
    if (*first)
      *first = false;
    else
      putchar(' ');

    printf("%s", name);
  }
}
#endif

static const char* get_battery_flag_string(unsigned flag)
{
  const char* name;

  switch (flag)
  {
    case BATTERY_FLAG_HIGH:        name = "High";       break;
    case BATTERY_FLAG_LOW:         name = "Low";        break;
    case BATTERY_FLAG_CRITICAL:    name = "Critical";   break;
    case BATTERY_FLAG_CHARGING:    name = "Charging";   break;
    case BATTERY_FLAG_NO_BATTERY:  name = "NoBattery";  break;

    default: name = "Unknown"; break;
  }

  return name;
}

static const char* get_ACLineStatus_string(unsigned ACLineStatus)
{
  const char* status;

  switch (ACLineStatus)
  {
    case AC_LINE_OFFLINE:       status = "Offline";       break;
    case AC_LINE_ONLINE:        status = "Online";        break;
    case AC_LINE_BACKUP_POWER:  status = "Backup Power";  break;
    case AC_LINE_UNKNOWN:       status = "Unknown";       break;
    default:                    status = "Invalid";       break;
  }

  return status;
}

void print_battery_status(const char* name, unsigned flag, unsigned lifePercent, unsigned lifeTime, unsigned fullLifeTime)
{
  printf(
      "\nStatus for %s battery\n"
      "=========================\n"
      "Flag:          %i (%s)\n"
      ,
      name,
      flag,
      get_battery_flag_string(flag));

  printf("LifePercent:   ");
  if (BATTERY_PERCENTAGE_UNKNOWN == lifePercent)
    printf("Unknown\n");
  else
    printf("%i%%\n", lifePercent);

  printf("LifeTime:      ");
  if (BATTERY_LIFE_UNKNOWN == lifeTime)
    printf("Unknown\n");
  else
    printf("%i\n", lifeTime);

  printf("FullLifeTime:  ");
  if (BATTERY_LIFE_UNKNOWN == fullLifeTime)
    printf("Unknown\n");
  else
    printf("%i\n", fullLifeTime);


}

int main(int argc, char** argv)
{
  int result = 1;
  RapiConnection* connection = NULL;
  HRESULT hr;
  CEOSVERSIONINFO version;
  SYSTEM_INFO system;
  SYSTEM_POWER_STATUS_EX power;
  STORE_INFORMATION store;
  DWORD storage_pages = 0, ram_pages = 0, page_size = 0;

  if (!handle_parameters(argc, argv))
    goto exit;

  if ((connection = rapi_connection_from_path(devpath)) == NULL)
  {
    fprintf(stderr, "%s: Could not find configuration at path '%s'\n", 
        argv[0],
        devpath?devpath:"(Default)");
    goto exit;
  }
  rapi_connection_select(connection);
  hr = CeRapiInit();

  if (FAILED(hr))
  {
    fprintf(stderr, "%s: Unable to initialize RAPI: %s\n", 
        argv[0],
        synce_strerror(hr));
    goto exit;
  }

  /*
   * Version
   */

  memset(&version, 0, sizeof(version));
  version.dwOSVersionInfoSize = sizeof(version);

  if (CeGetVersionEx(&version))
  {
    char *details = wstr_to_current(version.szCSDVersion);
    char *platform = NULL;

    if (VER_PLATFORM_WIN32_CE == version.dwPlatformId)
      platform = "(Windows CE)";

    printf(
        "Version\n"
        "=======\n"
        "Version:    %i.%i.%i (%s)\n"
        "Platform:   %i %s\n"
        "Details:    \"%s\"\n"
        "\n"
        ,
        version.dwMajorVersion,
        version.dwMinorVersion,
        version.dwBuildNumber,
        version_string(&version),
        version.dwPlatformId,
        platform ? platform : "",
        details
        );

    wstr_free_string(details);
  }
  else
  {
    fprintf(stderr, "%s: Failed to get version information: %s\n", 
        argv[0],
        synce_strerror(CeGetLastError()));
  }

  /*
   * System
   */

  memset(&system, 0, sizeof(system));

  CeGetSystemInfo(&system);
  {
    printf(
        "System\n"
        "======\n"
        "Processor architecture: %i (%s)\n"
        "Processor type:         %i (%s)\n"
        "Page size:              0x%x\n"
        "\n"
        ,
        system.wProcessorArchitecture,
        (system.wProcessorArchitecture < PROCESSOR_ARCHITECTURE_COUNT) ?
        architecture[system.wProcessorArchitecture] : "Unknown",
        system.dwProcessorType,
        processor(system.dwProcessorType),
        system.dwAllocationGranularity

        );
  }

  /*
   * Power
   */

  memset(&power, 0, sizeof(SYSTEM_POWER_STATUS_EX));

  if (CeGetSystemPowerStatusEx(&power, false))
  {
    printf(
        "Power\n"
        "=====\n"
        );

    printf("ACLineStatus: %02x (%s)\n", 
        power.ACLineStatus, get_ACLineStatus_string(power.ACLineStatus));

    print_battery_status("main", power.BatteryFlag, power.BatteryLifePercent,
        power.BatteryLifeTime, power.BatteryFullLifeTime);

    print_battery_status("backup", power.BackupBatteryFlag,
        power.BackupBatteryLifePercent, power.BackupBatteryLifeTime,
        power.BackupBatteryFullLifeTime);

    printf("\n");

  }
  else
  {
    fprintf(stderr, "%s: Failed to get battery status: %s\n", 
        argv[0],
        synce_strerror(CeGetLastError()));
  }

  /*
   * Store
   */
  memset(&store, 0, sizeof(store));

  if (CeGetStoreInformation(&store))
  {
    printf(
        "Store\n"
        "=====\n"
        "Store size: %i bytes (%i megabytes)\n"
        "Free space: %i bytes (%i megabytes)\n"
        "\n"
        ,
        store.dwStoreSize, store.dwStoreSize / (1024*1024),
        store.dwFreeSize,  store.dwFreeSize  / (1024*1024) 
        );
  }
  else
  {
    fprintf(stderr, "%s: Failed to get store information: %s\n", 
        argv[0],
        synce_strerror(CeGetLastError()));
  }

  if (CeGetSystemMemoryDivision(&storage_pages, &ram_pages, &page_size))
  {
    printf(
        "Memory for storage: %i bytes (%i megabytes)\n"
        "Memory for RAM:     %i bytes (%i megabytes)\n"
        "\n",
        storage_pages * page_size, storage_pages * page_size / (1024*1024),
        ram_pages     * page_size, ram_pages     * page_size / (1024*1024));
  }

  result = 0;

exit:
  CeRapiUninit();
  rapi_connection_destroy(connection);
  return result;
}
